src/entities.py

```diff
@@ -0,0 +1,129 @@
+
+# Ensure we are only in src dir
+SRCDIR = Path(__file__).resolve().parent
+os.chdir(SRCDIR)
+
+# Declare File content strings
+C_FILE_HEAD = """ /*
+ * Generated file - do not edit directly.
+ * This file was generated from:
+ *     https://html.spec.whatwg.org/entities.json
+ * by means of entities.py in the libgd src dir
+ */
+
+#include "entities.h"
+
+const struct entities_s entities[NR_OF_ENTITIES] = {
+"""
+H_FILE_HEAD = """ /*
+ * Generated file - do not edit directly.
+ * This file was generated from:
+ *     https://html.spec.whatwg.org/entities.json
+ * by means of entities.py in the libgd src dir
+ */
+
+#ifndef GD_ENTITIES_H
+#define GD_ENTITIES_H 1
+
+#include <stdlib.h>
+#include <stdint.h>
+
+"""
+H_FILE_TAIL = """
+#define ENTITY_HEX_LENGTH_MAX 10
+#define ENTITY_DEC_LENGTH_MAX 10
+
+// html entity strings are entity prefix + string + suffix limited
+// hex and dec should be limited to current unicode spec + entity prefix + suffix
+
+struct entities_s {
+                  	char *name;
+                        uint32_t unicode_a;
+                        uint32_t unicode_b;
+                };
+extern const struct entities_s entities[NR_OF_ENTITIES];
+
+#endif
+"""
+
+# Check for entities.json and fetch if needed
+if not os.path.exists("entities.json"):
+    print("entities.json not found, attempting to download...")
+    with open("entities.json", "wb") as file_json:
+        file_json.write(urllib.request.urlopen("https://html.spec.whatwg.org/entities.json").read())
+        print("\tsuccessful")
+
+# Load json obj
+with open("entities.json", "rb") as file_json:
+    entities = json.load(file_json)
+
+# Initialize some objects
+curline = 1
+name_matcher = re.compile(r"&\S+;")
+
+# Sum of total matching entities
+total_entities = sum(1 if name_matcher.match(key) else 0 for key in entities)
+
+# Find longest entity
+len_name_max = max(len(key) if name_matcher.match(key) else 0 for key in entities)
+
+# Write entities.h file
+with open("entities.h", mode="w") as file_ent_h:
+    file_ent_h.write(H_FILE_HEAD)
+    file_ent_h.write(f"#define NR_OF_ENTITIES {total_entities}\n");
+    file_ent_h.write(f"#define ENTITY_NAME_LENGTH_MAX {len_name_max}");
+    file_ent_h.write(H_FILE_TAIL)
+
+# Write entities.c file
+with open("entities.c", mode="w") as file_ent_c:
+    file_ent_c.write(C_FILE_HEAD)
+    # Write json entities to struct
+    for key in entities:
+        if name_matcher.match(key):
+            string = "\t" + "{\"" + key.replace("&", "") + "\", "
+            string = string.replace(";", "")
+            codepoints = entities[key]["codepoints"]
+            string = string + str(codepoints[0]) + ", "
+            if len(codepoints) > 1:
+                string = string + str(codepoints[1]) + "}"
+            else:
+                string = string + "0}"
+            if len(codepoints) > 2:
+                print(f"Warning: entity with >2 codepoints detected")
+            if curline == total_entities:
```

Comment:
just omit this check.  dangling commas are fine.  just write:

`string += ",\n"`